core/lib/dal/src/via_votes_dal.rs

```diff
@@ -0,0 +1,131 @@
+        &mut self,
+        l1_batch_number: u32,
+        tx_id: H256,
+    ) -> DalResult<()> {
+        sqlx::query!(
+            r#"
+            INSERT INTO
+                via_votable_transactions (l1_batch_number, tx_id)
+            VALUES
+                ($1, $2)
+            ON CONFLICT (l1_batch_number, tx_id) DO NOTHING
+            "#,
+            l1_batch_number as i64,
+            tx_id.as_bytes()
+        )
+        .instrument("insert_votable_transaction")
+        .fetch_optional(self.storage)
+        .await?;
+
+        Ok(())
+    }
+
+    /// Inserts a new vote row in `via_votes`.
+    /// Now requires `l1_batch_number` as part of the primary key / FK.
+    pub async fn insert_vote(
+        &mut self,
+        l1_batch_number: u32,
+        tx_id: H256,
+        verifier_address: &str,
+        vote: bool,
+    ) -> DalResult<()> {
+        sqlx::query!(
+            r#"
+            INSERT INTO
+                via_votes (l1_batch_number, tx_id, verifier_address, vote)
+            VALUES
+                ($1, $2, $3, $4)
+            ON CONFLICT (l1_batch_number, tx_id, verifier_address) DO NOTHING
+            "#,
+            l1_batch_number as i32,
+            tx_id.as_bytes(),
+            verifier_address,
+            vote
+        )
+        .instrument("insert_vote")
+        .fetch_optional(self.storage)
+        .await?;
+
+        Ok(())
+    }
+
+    /// Returns (ok_votes, total_votes) for the given `(l1_batch_number, tx_id)`.
+    /// Must also filter on `l1_batch_number`.
+    pub async fn get_vote_count(
+        &mut self,
+        l1_batch_number: u32,
+        tx_id: H256,
+    ) -> DalResult<(i64, i64)> {
+        let row = sqlx::query!(
+            r#"
+            SELECT
+                COUNT(*) FILTER (
+                    WHERE
+                        vote = TRUE
+                ) AS ok_votes,
+                COUNT(*) AS total_votes
+            FROM
+                via_votes
+            WHERE
+                l1_batch_number = $1
+                AND tx_id = $2
+            "#,
+            l1_batch_number as i32,
+            tx_id.as_bytes()
+        )
+        .instrument("get_vote_count")
+        .fetch_one(self.storage)
+        .await?;
+
+        let ok_votes = row.ok_votes.unwrap_or(0);
+        let total_votes = row.total_votes.unwrap_or(0);
+        Ok((ok_votes, total_votes))
+    }
+
+    /// Marks the transaction as finalized if #ok_votes / #total_votes > threshold.
+    /// Must use `(l1_batch_number, tx_id)` in both vote counting and the UPDATE statement.
+    pub async fn finalize_transaction_if_needed(
+        &mut self,
+        l1_batch_number: u32,
+        tx_id: H256,
+        threshold: f64,
+    ) -> DalResult<bool> {
+        let (ok_votes, total_votes) = self.get_vote_count(l1_batch_number, tx_id).await?;
+        let is_above_threshold =
+            total_votes > 0 && (ok_votes as f64) / (total_votes as f64) > threshold;
```

Comment:
We probably need `>=` here and something like `is_threshold_reached` (for example if we need 4-of-5 and threshold is 0.8 it needs to pass)